### PR TITLE
week9

### PR DESCRIPTION
**Pull request by NatachaKey**

### CODE REVIEW

File: middleware/authentication.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  const token = authHeader.split(' ')[1];
+  try {
+    const payload = jwt.verify(token, process.env.JWT_SECRET);
+    //attach the user to the job routes
+    req.user = { userId: payload.userId, name: payload.name };
+    next()
+  } catch (error) {
+    throw new UnauthenticatedError('Authentication invalid')
+  }
+};
```
Comment (akosasante) [inline middleware/authentication.js:20]: To help answer your question I'll try to break down what's happening on each line of this function:

*   Line 7: Check what headers were passed to the server in the request. Specifically, save the value of the header with the name `"Authorization"` into the variable `authHeader`
*   Lines 8-10: We expect the value to look like: `"Bearer <the user's JWT token>"`. So here we check if we didn't get any value at all (the request didn't have a header with the name `"Authorization"`, or the value of that header was empty. _Or,_ if it did have a value but that value didn't start with the text `"Bearer"`, then we can just stop right there and throw an authentication error (which due to our `error-handler` middleware returns a 401 status response)
*   Line 11: We only care about the actual token, so we split the text `"Bearer <the user's JWT token>"`, which gives us an array like `["Bearer", <the user's JWT token>]` and then grab the second item in that aray by using the `[1]` square brackets to access that index
*   Line 13: We use the `jsonwebtoken` library to decode the string token that was in the authorization header.   
      
    Earlier, when the user first registered using the `/api/v1/auth/register` endpoint, or logged in with `/api/v1/auth/login` we called an instance method on the `UserSchema` model called `createJWT`:  
    [https://github.com/NatachaKey/06-jobs-api/pull/1/files#diff-dd22ac78fcf815fae5fdb7a59f4a9f6db8b31971d0e7bca671dd6c3025d47f13R34-R42](https://github.com/NatachaKey/06-jobs-api/pull/1/files#diff-dd22ac78fcf815fae5fdb7a59f4a9f6db8b31971d0e7bca671dd6c3025d47f13R34-R42)  
      
    In that function we told the `jsonwebtoken` library to take the object `{ userId: this._id, name: this.name }` and encode it as a JSON Web Token using the `JWT_SECRET` in our `.env` file. That means that only someone who uses `JWT_SECRET` can decode the JSON Web Token string back into the object. Check out the following recording of me in node encoding and decoding some secrets.  
      
    https://asciinema.org/a/sXnlExFMm3CpAIx6IKX5BsTRm  
      
    Now here, when someone is trying to use any of the `/api/v1/jobs` routes, we want to make sure we know what user is making the request. So we take the JSON Web Token string from the request headers and decode it using the same `JWT_SECRET` and get back an object that has a `userId` field and a `name` field. We store that object back in the request on a field we call `user`. That way, all the routes using this middleware, can check which user is making the request by just looking at the `req.user` field.  
     
*   Lines 17-19: All of the work above was done in a `try/catch` block; so if `jwt.verify` throws an error, we want to make sure to catch it and convert that to one of the errors that our `error-handler` middleware knows how to deal with.

Comment (akosasante) [reply middleware/authentication.js:20]: Here's an example of me using https://jwt.io/ to decode the token that I created in the recorded video; you can see that even without the secret, the website can fully decode the token back into its original payload. So we never want to put sensitive info like passwords into a JWT:

**Review by akosasante — APPROVED** [review]
Everything looks good!

Feel free to drop a comment if you have any more questions about the work this week :smile:

File: middleware/authentication.js (hunk 1 above)
Comment (NatachaKey) [reply middleware/authentication.js:20]: @akosasante thank you so much, Akos, now everything looks much more clear now!

**Pull request merged**